### ISSUE TITLE
[Bug] Hihats sounding while they shouldn't in BPS Tetris [sf#511]

### ISSUE THREAD

**Issue by openMSX-import** [openMSX/openMSX#903]
**Reported by manuelbi on 2014-01-07 20:32 UTC**
When playing BPS Tetris with the Technotris tune, you can hear weird Hi-Hat sounds when the drums kick in. This wasn't happening with 0.9.1.

More info when I get it...

**Comment by openMSX-import** [comment]
**Commented by manuelbi on 2014-01-07 20:34 UTC**
The issue is not present when using the alternate core.

**Comment by openMSX-import** [comment]
**Commented by manuelbi on 2014-01-07 20:53 UTC**
Bisecting on the Okazaki core gives me this result:
323023f877823ca53453a58350c1db84977780b9 is the first bad commit
commit 323023f877823ca53453a58350c1db84977780b9
Author: Wouter Vermaelen m9710797@users.sourceforge.net
Date:   Mon Feb 18 16:39:28 2013 +0000

```
Fixed bug [3603184] Changing to rhythm mode disables channels on Okazaki core

For more detailed problem description see:
   https://sourceforge.net/tracker/?func=detail&atid=421861&aid=3603184&group_id=38274

The Okazaki core showed the problem, the Burczynski core was correct (at least
as far as I can tell it behaved the same as the real machine for this test).
This made debugging this issue a lot easier.

The difference between the two cores was that Okazaki forced all envelope
states to the 'FINISH' state when enabling rhythm mode. Burczynski left this
state unchanged. Thus fixed by also not changing this state in Okazaki.

After this change, the Okazaki core sometimes had a 'hanging' drum channel
(e.g. after commenting out line 91 in the test program, ch8.mod (the 'TOM'
drum) would keep on producing an annoying beep). This was caused by not
properly recalculating all internal state after the 'patch' parameters were
changed (more specifically 'eg_phase_max' was not recalculated after 'patch.SL'
was changed). It's possible that this 2nd bug also surfaced in other situations
(e.g. when switching instrument while another was still playing).
```

**Comment by openMSX-import** [comment]
**Updated by manuelbi on 2014-01-11 13:54 UTC**
- **status**: open --> closed-fixed
- **assigned_to**: Wouter Vermaelen

**Comment by openMSX-import** [comment]
**Commented by manuelbi on 2014-01-11 13:54 UTC**
Fixed by wouter with commit 0a3c6aad445734f6678a157ee91a562e4b129110
Thanks!
